### PR TITLE
Merged wpt-dbw branches and tl_detector.

### PR DESCRIPTION
**Pull request by khatiba**
I did my best to merge the work done so far for the DBW node and part of the waypoint updater. I had a pretty good implementation of the controller, it has steady throttle and braking, I scanned William's repo and it looks about right. I took most of Xiang's waypoint updater code but left out the traffic light detection for now. This branch just drives the car around at constant velocity and ignores lights.

### CODE REVIEW

**Review by xiangjiang — COMMENTED** [review]
~~This works great on the Highway. However, when I tried it on test lot, it still stuck after about 3/4 of a circle.~~
~~To try it on the test lot, I modified the launch/site.launch to add the simulator bridge (which is present in launch/styx.launch). Then use roslaunch launch/site.launch with the test lot.~~
~~The problem is steering is not enough when the circle is small and angular movement is large.~~
~~Naive solution here is to increase steering no matter what by about 30%.~~
~~Then the car can do infinite loop in test lot.~~

~~Let me know your thoughts.~~

Sorry I think after removing these in the pull request everything works perfectly in both scenarios.
`            # TODO: Temporarily set speed to optimize controller gains`
`            for i in range(len(final_waypoints)):`
`                self.set_waypoint_velocity(final_waypoints, i, 20.0)`

**Comment by khatiba** [comment]
Ah, yes, I didn't test the control gains with slower speeds in the test lot, I'll try that out with slower speeds, ~~maybe 25mph or so.~~ 10mph

**Comment by khatiba** [comment]
I tested on the church lot and highway but with the speed limits set in the launch files, it performed okay, but I feel like just a small increase in steering would be better. I tried 10,20,30% increases and it was quite a bit better.

**Review by haribalanr — APPROVED** [review]

**Comment by xiangjiang** [comment]
Stopping function has very high quality now.
I like the setting overshoot waypoint to target speed -1.0.
There is probably on minor problem I left in the resuming part.
`target_speed = 100/9`
I hardcoded it. I think we can get is from rosparam rather than hardcoding it.
Otherwise, if we include Hari's tl_classifer.py, I think it worth submitting.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 1]
```diff
@@ -21,36 +22,139 @@
...
+            # rospy.loginfo("Red light ahead, set velocity to stop")
+
+            wps_to_light = light_wp - self.wp_current_start
+            for i in range(wps_to_light):
+                delta = self.distance(self.all_wps.waypoints, self.wp_current_start, self.wp_current_start + i)
+                if dist_to_light > 0:
+                    new_speed = -1*(current_speed/dist_to_light) * delta + current_speed
+                    self.set_waypoint_velocity(self.all_wps.waypoints, i + self.wp_current_start, new_speed)
+                else:
+                    self.set_waypoint_velocity(self.all_wps.waypoints, j + self.wp_current_start, -1.0)
```
Comment (xiangjiang) [inline ros/src/waypoint_updater/waypoint_updater.py]: I think there is a typo here, j should be i

Comment (khatiba) [reply ros/src/waypoint_updater/waypoint_updater.py]: Good catch! I guess that actually never got called yet. Fixed now.

**Comment by khatiba** [comment]
I'll check out the resuming part now.

**Comment by khatiba** [comment]
Xiang, I found some of your commented out code and added my own changes so that the car will linearly increase velocity from a stop, it's basically the reverse of the braking loop. This also respects the speed limit set in ros params. I also forgot about the low pass filter, so I added that to smooth the throttle when accelerating.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 2]
```diff
@@ -29,10 +29,11 @@ class WaypointUpdater(object):
     def __init__(self):
         rospy.init_node('waypoint_updater')
 
+        self.speed_limit = rospy.get_param('/waypoint_loader/velocity', 15) * 0.45 # MPH to m/s
```
Comment (xiangjiang) [inline ros/src/waypoint_updater/waypoint_updater.py]: I think waypoint loader velocity param is Kilometer per hour.
40 KPH = 100/9 m/s = 24.85 MPH
The multiplier is then 5/18 = 0.277778

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 3]
```diff
@@ -128,33 +130,35 @@ def traffic_cb(self, msg):
...
                 self.all_wps = self.original_wps
-                for i in range(resuming_in_action):
-                    rospy.loginfo("i is %s", i)
-                    #self.set_waypoint_velocity(self.all_wps.waypoints,i+self.wp_current_start,own_speed+(target_speed-own_speed)*(i+1)/(resuming_in_action))
-                    self.set_waypoint_velocity(self.all_wps.waypoints,i+self.wp_current_start,target_speed)
-                for i in range(110):
-                    self.set_waypoint_velocity(self.all_wps.waypoints,i+self.wp_current_start+resuming_in_action,target_speed)
+                num_waypoints = len(self.all_wps.waypoints)
+                for j in range(i, num_waypoints - num_resume_wps - 1):
+                    self.set_waypoint_velocity(self.all_wps.waypoints, (j+self.wp_current_start) % num_waypoints, self.speed_limit)
```
Comment (xiangjiang) [inline ros/src/waypoint_updater/waypoint_updater.py]: C.f. the previous comment, now we don't need to set all the way to the end with speed_limit. Just need to make sure we change previously slowed velocity back.

**Comment by xiangjiang** [comment]
I have now included tl_detector.py that I posted in slack. It is not calling get_light_state. 
If Hari is going to add his light classification, please uncomment line 210 and comment line 211, then add your NN with the steps Hari described in slack.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 4]
```diff
@@ -21,36 +22,143 @@
...
+        num_waypoints = len(self.all_wps.waypoints)
+
+        closest_point = 0
+        max_distance = float('inf')
+        dl = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2  + (a.z-b.z)**2)
+
+        for i in range(self.last_closest_waypoint, self.last_closest_waypoint + LOOKAHEAD_WPS):# len(self.all_wps.waypoints)):
+            current_distance = dl(pos, self.all_wps.waypoints[i % num_waypoints].pose.pose.position)
+            if current_distance < max_distance:
+                closest_point = i
```
Comment (xiangjiang) [inline ros/src/waypoint_updater/waypoint_updater.py]: closest_point = i % num_waypoints
Found this out while trying site.launch

Comment (khatiba) [reply ros/src/waypoint_updater/waypoint_updater.py]: ah! good catch

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (khatiba) [reply ros/src/waypoint_updater/waypoint_updater.py]: yes your right, that makes sense.

**Pull request merged**